### PR TITLE
RTPS AckNack is not scalable

### PR DESCRIPTION
**Pull request by jrw972**
Problem
-------

A reliable RtpsReader loops over all writers when generating AckNacks.

Solution
--------

Keep track of two sets: writers that need a nack due to missing data
and writers that just need an ack.

This also adds support for sending HeartBeatFrag messages.

### CODE REVIEW

File: dds/DCPS/RTPS/MessageTypes.h [hunk 1]
```diff
@@ -22,6 +22,7 @@ namespace OpenDDS {
     const ACE_CDR::UShort RTPSHDR_SZ = 20, // size of RTPS Message Header
       SMHDR_SZ = 4, // size of SubmessageHeader
       HEARTBEAT_SZ = 28, // size (octetsToNextHeader) of HeartBeatSubmessage
+      HEARTBEAT_FRAG_SZ = 24, // size (octetsToNextHeader) of HeartBeatFragSubmessage
```
Comment (jrw972) [inline dds/DCPS/RTPS/MessageTypes.h:25]: Remove this.

File: dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp [hunk 2]
```diff
@@ -3599,24 +3669,35 @@ RtpsUdpDataLink::RtpsWriter::gather_heartbeats(OPENDDS_VECTOR(TransportQueueElem
...
 
-      const SequenceNumber sn = expected_max_sn(ri);
-      meta_submessage.dst_guid_ = ri->id_;
-      meta_submessage.sm_.heartbeat_sm().readerId = ri->id_.entityId;
-      meta_submessage.sm_.heartbeat_sm().lastSN.low = sn.getLow();
-      meta_submessage.sm_.heartbeat_sm().lastSN.high = sn.getHigh();
-      meta_submessages.push_back(meta_submessage);
+      SequenceNumber last_fragment;
+      // TODO:  Send one or both?
+      if (!send_buff_->is_fragmented(ri->cur_cumulative_ack_, last_fragment)) {
```
Comment (jrw972) [inline dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp:3674]: ack - 1

File: dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp [hunk 3]
```diff
@@ -3599,24 +3669,35 @@ RtpsUdpDataLink::RtpsWriter::gather_heartbeats(OPENDDS_VECTOR(TransportQueueElem
...
+        const SequenceNumber sn = expected_max_sn(ri);
+        heartbeat_meta_submessage.dst_guid_ = ri->id_;
+        heartbeat_meta_submessage.sm_.heartbeat_sm().readerId = ri->id_.entityId;
+        heartbeat_meta_submessage.sm_.heartbeat_sm().lastSN.low = sn.getLow();
+        heartbeat_meta_submessage.sm_.heartbeat_sm().lastSN.high = sn.getHigh();
+        meta_submessages.push_back(heartbeat_meta_submessage);
+      } else {
```
Comment (jrw972) [inline dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp:3681]: Don't send HeartBeatFrag.

File: dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp [hunk 4]
```diff
@@ -2060,59 +2109,55 @@ RtpsUdpDataLink::RtpsReader::gather_ack_nacks_i(MetaSubmessageVec& meta_submessa
...
-
-      EntityId_t reader_id = id_.entityId, writer_id = wi->first.entityId;
-
-      MetaSubmessage meta_submessage(id_, wi->first);
-
       AckNackSubmessage acknack = {
         {ACKNACK,
-         CORBA::Octet(FLAG_E | (is_final ? FLAG_F : 0)),
-         0 /*length*/},
-        id_.entityId,
```
Comment (jrw972) [inline dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp:2094]: Restore logic that clears bitmap where entries are known fragments.

File: dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp [hunk 5]
```diff
@@ -2009,36 +2061,33 @@ RtpsUdpDataLink::RtpsReader::gather_ack_nacks_i(MetaSubmessageVec& meta_submessa
...
+  // These writers need a nack.
+  for (WriterInfoSet::const_iterator pos = writers_expecting_nack_.begin(), limit = writers_expecting_nack_.end();
+       pos != limit; ++pos) {
+    const WriterInfo_rch& info = *pos;
+
+    const EntityId_t reader_id = id_.entityId;
+    const EntityId_t writer_id = info->id_.entityId;
+    MetaSubmessage meta_submessage(id_, info->id_);
+
+    // TODO: Send one or both?
```
Comment (jrw972) [inline dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp:2073]: Send both.

File: dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp [hunk 6]
```diff
@@ -1981,21 +2004,50 @@ RtpsUdpDataLink::RtpsReader::writer_count() const
...
+  // 1. sequence #s in the reception gaps that we have partially received
+  OPENDDS_VECTOR(SequenceRange) missing = info->recvd_.missing_sequence_ranges();
+  for (size_t i = 0; i < missing.size(); ++i) {
+    if (link->receive_strategy()->has_fragments(missing[i], info->id_)) {
+      return true;
+    }
+  }
+  // 1b. larger than the last received seq# but less than the heartbeat.lastSN
+  if (!info->recvd_.empty()) {
+    const SequenceRange range(info->recvd_.high(), info->hb_range_.second);
```
Comment (jrw972) [inline dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp:2027]: use recvd_ cur_cumulative_ack

**Review by mitza-oci — APPROVED** [review]

**Pull request merged**